kotlin/src/main/kotlin/org/jdbi/v3/core/kotlin/KotlinMapper.kt

```diff
@@ -11,108 +11,190 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
+
 package org.jdbi.v3.core.kotlin
 
-import org.jdbi.v3.core.mapper.ColumnMapper
+import org.jdbi.v3.core.mapper.Nested
 import org.jdbi.v3.core.mapper.RowMapper
+import org.jdbi.v3.core.mapper.SingleColumnMapper
 import org.jdbi.v3.core.mapper.reflect.ColumnName
 import org.jdbi.v3.core.mapper.reflect.ColumnNameMatcher
+import org.jdbi.v3.core.mapper.reflect.ReflectionMapperUtil.*
 import org.jdbi.v3.core.mapper.reflect.ReflectionMappers
 import org.jdbi.v3.core.statement.StatementContext
-import java.lang.reflect.InvocationTargetException
 import java.sql.ResultSet
-import java.sql.SQLException
-import java.util.*
 import java.util.concurrent.ConcurrentHashMap
-import kotlin.reflect.*
+import kotlin.reflect.KClass
+import kotlin.reflect.KFunction
+import kotlin.reflect.KMutableProperty1
+import kotlin.reflect.KParameter
 import kotlin.reflect.full.findAnnotation
 import kotlin.reflect.full.memberProperties
 import kotlin.reflect.full.primaryConstructor
 import kotlin.reflect.jvm.isAccessible
 import kotlin.reflect.jvm.javaField
 import kotlin.reflect.jvm.javaType
+import kotlin.reflect.jvm.jvmErasure
 
-private typealias ValueProvider = (r: ResultSet, c: StatementContext) -> Any?
-private val NULL_VALUE_PROVIDER: ValueProvider = { _, _ -> null }
-
-
-class KotlinMapper<C : Any>(clazz: Class<C>) : RowMapper<C> {
-
-    private val kClass: KClass<C> = clazz.kotlin
-
+class KotlinMapper(clazz: Class<*>, private val prefix: String = "") : RowMapper<Any> {
+    private val kClass: KClass<*> = clazz.kotlin
     private val constructor = findConstructor(kClass)
     private val constructorParameters = constructor.parameters
-    private val nullableMandatoryParameters = constructorParameters.filter { !it.isOptional && it.type.isMarkedNullable }
-    private val mutableProperties = kClass.memberProperties.map { it as? KMutableProperty1 }.filterNotNull()
+    private val memberProperties = kClass.memberProperties
```

Comment:
Are `constructorParameters` and `memberProperties` mutually exclusive? If not, then `memberProperties` should probably be deduped.

Reply:
As far as I know data classes will not expose any member properties through reflection, but I might be wrong. I'll make sure and get back to you on this.

Reply:
You were right, member properties will show up for all `val`/`var` constructor parameters (in regular and data classes). I'll figure something out to remove dupes.